src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,225 +138,134 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var declarationStatement = (StatementSyntax)declarator.Parent.Parent;
+            // Check if the as operand is ever written up to the point of null check.
+            //
+            //      var s = field as string;
+            //      field = null;
+            //      if (s != null) { ... }
+            //
+            // It's no longer safe to use pattern-matching because 'field is string s' would never be true.
+            var asOperand = semanticModel.GetSymbolInfo(asExpression.Left, cancellationToken).Symbol;
+            var symbolName = asOperand?.Name;
+            if (symbolName != null)
+            {
+                var localStatementStart = localStatement.SpanStart;
+                var comparisonSpanStart = comparison.SpanStart;
+
+                foreach (var descendentNode in enclosingBlock.DescendantNodes())
+                {
+                    var descendentNodeSpanStart = descendentNode.SpanStart;
+                    if (descendentNodeSpanStart <= localStatementStart)
+                    {
+                        continue;
+                    }
+
+                    if (descendentNodeSpanStart >= comparisonSpanStart)
+                    {
+                        break;
+                    }
 
-            // If we convert this to 'if (o is Type x)' then 'x' will not be definitely assigned
-            // in the Else branch of the IfStatement, or after the IfStatement. Make sure
-            // that doesn't cause definite assignment issues.
-            if (!CheckDefiniteAssignment(semanticModel, localSymbol,
-                    declarationStatement, targetStatement, parentBlock, isNegativeNullCheck))
+                    if (descendentNode.IsKind(SyntaxKind.IdentifierName, out IdentifierNameSyntax identifierName) &&
+                        identifierName.Identifier.ValueText == symbolName &&
+                        asOperand.Equals(semanticModel.GetSymbolInfo(identifierName, cancellationToken).Symbol) &&
+                        identifierName.IsOnlyWrittenTo())
```

Comment:
>IsOnlyWrittenTo [](start = 39, length = 15)

Why "only"?
I would think that a read-write access (like a compound assignment) should count as a write for the purpose of this analysis.